### PR TITLE
Querying ACSets

### PR DESCRIPTION
**Pull request by olynch**
I accidentally deleted the branch for the previous PR, #340, so this continues that PR. Also, this includes queries. Given that the ACSetView structure still needs review, I figure that it makes sense to have both the ACSetView structure and the queries on the same PR, because changes to one will probably necessitate changes in the other.

### CODE REVIEW

**Comment by epatters** [comment]
Writing this down here before I forget: `ACSetView` should implement the [Tables.jl](https://github.com/JuliaData/Tables.jl) interface. This won't add any additional dependency since Catlab already indirectly depends on Tables.jl through StructArrays.jl.

**Comment by olynch** [comment]
OK! Will do.

**Comment by epatters** [comment]
Great, thanks! And as of #350, Catlab will depend explicitly on Tables.jl.

**Comment by olynch** [comment]
This is ready for review @epatters, though I already know I need to add a test for ACSetViewMorphism.

File: src/categorical_algebra/ACSetViews.jl [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  print(io, "ACSetView of object $P")
+  println(io)
+  table = tables(backing(av))[P][parts(av)]
+  if !(eltype(table) <: NamedTuple{(),Tuple{}} || isempty(table))
+    pretty_table(io, table, nosubheader=true,
+                 show_row_number=true, row_number_column_title=string(P))
+  end
+end
+
+# Tables.jl interface
```
Comment (epatters) [inline src/categorical_algebra/ACSetViews.jl:90]: It looks like most of the Tables.jl interface is not covered by tests. Would be good to add a simple test for each function.

File: src/categorical_algebra/ACSetViews.jl [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+# Macros for Querying/filtering
+###############################
+
+""" This macro takes an expression using attributes of the object selected by the ACSetView,
+and returns an array of the result of that expression applied to the attribute of each part
+in the ACSetView.
+
+Example: @compute_prop(edges_table, 2 * src.weight)
+
+Eventually, this should be superceded by Query.jl or AlgebraicRelations.jl functionality, so I have
```
Comment (epatters) [inline src/categorical_algebra/ACSetViews.jl:117]: Lol at "purposely made the name a bit obtuse and awkward"!

The ability to use compound attributes is nice. IMO, if you think this is useful (it seems useful to me!), you should give it a worthy name.

File: src/categorical_algebra/ACSetViews.jl [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    Expr(:., x, y) => [flatten_dot_expr(x)..., y.value]
+    _ => error("unsupported syntax in @query")
+  end
+end
+
+function flatten_dot_expr(ex::Symbol)
+  [ex]
+end
+
+""" This macro takes in a similar expression to @compute prop, except that the expression
```
Comment (epatters) [inline src/categorical_algebra/ACSetViews.jl:150]: `@compute prop` -> ```[`@compute_prop`](@ref)```

**Review by epatters — COMMENTED** [review]
Thanks Owen, this looks really nice! Minor comments above. The main thing is to finish the tests, then this should be ready to merge.

File: src/categorical_algebra/ACSetViews.jl (hunk 2 above)
Comment (olynch) [reply src/categorical_algebra/ACSetViews.jl:117]: I think this function is useful, but I think that it should be replaced by something more useful. And actually, I think `@compute_prop` is not so bad.

**Comment by epatters** [comment]
Thanks, I'll merge this!

Not sure what is going on with Codecov. It is claiming that lots of things you clearly have tests for are uncovered. Hopefully this problem does not persist.

File: src/categorical_algebra/CSetDataStructures.jl [hunk 4]
```diff
@@ -871,4 +873,39 @@ function init_acset(T::Type{<:ACSet{CD,AD,Ts}},body) where {CD <: CatDesc, AD <:
   code
 end
 
+""" Map over a data type, in the style of Haskell functors
+"""
+
+Base.map(B::Type, f::Function, D::Symbol, acs::ACSet) = _map(B, f,Val{D},acs)
```
Comment (epatters) [inline src/categorical_algebra/CSetDataStructures.jl:879]: Did this sneak in from #352? It's not supposed to be part of this PR, right?

**Review by epatters — COMMENTED** [review]
I just realized why the code coverage is so low: the tests aren't running on CI because they aren't included in `test/categorical_algebra/CategoricalAlgebra.jl`! Also, another question below.

**Comment by olynch** [comment]
Oh whoops! I'll fix those two issues, I guess I must have mixed them up when merging or something, and I was running the tests through an ad-hoc test set so I completely forgot to add them!

**Comment by epatters** [comment]
Thanks, once the tests are passing, this should be ready to go.

**Comment by olynch** [comment]
Looks like this is one of the few times where the correct fix is to remove a test!

**Comment by olynch** [comment]
It's failing because it can't import Tables.jl... The tests run on my machine, so I think it might be a CI bug, especially because Tables is in the Project.toml.

**Comment by epatters** [comment]
Ah, it's failing because the tests have their own [Project.toml](https://github.com/AlgebraicJulia/Catlab.jl/blob/master/test/Project.toml), which is needed because the tests use various optional and extra dependencies that Catlab doesn't depend on.

**Comment by epatters** [comment]
Good news: Now that #358 is merged, Tables.jl is included in `tests/Project.toml`, so if you rebase, the tests will hopefully pass on CI.

Bad news: There's now a small merge conflict, which you can also fix by rebasing.

**Review by epatters — APPROVED** [review]
LGTM. Thanks Owen!

**Pull request merged**